login/pam.c

```diff
@@ -0,0 +1,272 @@
+  if (get_authcode(host_id, action, config->service_key, config->secret_key,
+                   authcode)) {
+    free(host_id);
+    free(action);
+    return failure(EXITCODE_PANIC, error_tag, "get-authcode");
+  }
+
+  char *url = NULL;
+  int url_len = 0;
+  if (request_url(config->service_key, config->service_key_id, public_key,
+                  host_id, action, /*prefix_tag=*/NULL,
+                  /*prefix_tag_len=*/0, &url, &url_len, error_tag)) {
+    free(host_id);
+    free(action);
+    return EXITCODE_PANIC;
+  }
+
+  free(host_id);
+  host_id = NULL;
+  free(action);
+  action = NULL;
+
+  char *input = NULL;
+  if (pam_prompt(pamh, PAM_PROMPT_ECHO_ON, &input,
+                 "%s%s\nObtain GLOME code from the above link: ",
+                 config->url_prefix, url)) {
+    return failure(EXITCODE_PANIC, error_tag, "pam-prompt");
+  }
+
+  free(url);
+  url = NULL;
+
+  int bytes_read = strlen(input);
+  if (config->options & INSECURE) {
+    pam_syslog(pamh, LOG_DEBUG, "user input: %s", input);
+  }
+
+  // Calculate the correct authcode.
+  char authcode_encoded[ENCODED_BUFSIZE(sizeof authcode)] = {0};
+  if (base64url_encode(authcode, sizeof authcode, (uint8_t *)authcode_encoded,
+                       sizeof authcode_encoded) == 0) {
+    free(input);
+    return failure(EXITCODE_PANIC, error_tag, "authcode-encode");
+  }
+  if (config->options & INSECURE) {
+    pam_syslog(pamh, LOG_DEBUG, "expect input: %s", authcode_encoded);
+  }
+
+  if (bytes_read < MIN_ENCODED_AUTHCODE_LEN) {
+    free(input);
+    return failure(EXITCODE_INVALID_INPUT_SIZE, error_tag, "authcode-length");
+  }
+  if (bytes_read > strlen(authcode_encoded)) {
+    free(input);
+    return failure(EXITCODE_INVALID_INPUT_SIZE, error_tag, "authcode-length");
+  }
+
+  // Since we use (relatively) short auth codes, sleep before confirming the
+  // result to prevent bruteforcing.
+  struct timespec delay;
+  delay.tv_sec = (time_t)config->auth_delay_sec;
+  delay.tv_nsec = 0;
+  if (nanosleep(&delay, NULL) != 0) {
+    free(input);
+    return failure(EXITCODE_INTERRUPTED, error_tag, "auth-delay");
+  }
+
+  // We can use insecure strncmp() because it's a *one-time* code.
+  if (strncmp(input, authcode_encoded, bytes_read) != 0) {
+    free(input);
+    return failure(EXITCODE_INVALID_AUTHCODE, error_tag, "authcode-invalid");
+  }
+
+  free(input);
+  return 0;
+}
+
+int pam_sm_authenticate(pam_handle_t *pamh, int flags, int argc,
+                        const char **argv) {
+  const char *error_tag = NULL;
+  login_config_t config = {0};
+  int rc = PAM_AUTH_ERR;
+
+  int r = parse_pam_args(pamh, argc, argv, &config);
+  if (r < 0) {
+    pam_syslog(pamh, LOG_ERR, "failed to parse pam module arguments (%d)", r);
+    return rc;
+  }
+
+  r = parse_config_file(&config);
+  if (r < 0) {
+    pam_syslog(pamh, LOG_ERR, "failed to read config file: %s (%d)",
+               config.config_path, r);
+    return rc;
+  }
```

Comment:
If I have a pam module config that specifies a custom service key and there happens to be a global config file with another service key, the global one would be chosen, wouldn't it? This may be unexpected.

Reply:
Yes, this behaviour is mentioned already in #33. Patch is now updated to prefer the parameter over the config file.